Add DataFrame reference to the user guide (#3067)
apache/arrow-datafusion: docs/source/user-guide/dataframe.md

<!---
  Licensed to the Apache Software Foundation (ASF) under one
  or more contributor license agreements.  See the NOTICE file
  distributed with this work for additional information
  regarding copyright ownership.  The ASF licenses this file
  to you under the Apache License, Version 2.0 (the
  "License"); you may not use this file except in compliance
  with the License.  You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing,
  software distributed under the License is distributed on an
  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
  KIND, either express or implied.  See the License for the
  specific language governing permissions and limitations
  under the License.
-->

> # DataFrame API

Reviewer: I wonder if it is important to mention somewhere that the computations are deferred until `collect()` is called? Maybe that is common across other dataframe implementations and can be assumed.

A DataFrame represents a logical set of rows with the same named columns, similar to a [Pandas DataFrame](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html) or
[Spark DataFrame](https://spark.apache.org/docs/latest/sql-programming-guide.html).

DataFrames are typically created by calling a method on
`SessionContext`, such as `read_csv`, and can then be modified
by calling the transformation methods, such as `filter`, `select`, `aggregate`, and `limit`
to build up a query definition.

The query can be executed by calling the `collect` method.

> The API is well documented at https://docs.rs/datafusion/latest/datafusion/dataframe/struct.DataFrame.html

Reviewer: ```suggestion
The API is well documented in the [API reference on docs.rs](https://docs.rs/datafusion/latest/datafusion/dataframe/struct.DataFrame.html)
```

The DataFrame struct is part of DataFusion's prelude and can be imported with the following statement.

```rust
use datafusion::prelude::*;
```

Here is a minimal example showing the execution of a query using the DataFrame API.

```rust
let ctx = SessionContext::new();
let df = ctx.read_csv("tests/example.csv", CsvReadOptions::new()).await?;
let df = df.filter(col("a").lt_eq(col("b")))?
           .aggregate(vec![col("a")], vec![min(col("b"))])?
           .limit(None, Some(100))?;

> let results = df.collect();

Reviewer: I wonder if using `show` would be a good idea in this example (and then we could include expected output as well):

```suggestion
// Print results
df.show();
```

https://github.com/apache/arrow-datafusion/blob/master/datafusion/core/src/dataframe.rs#L392-L408

| --------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| explain         | Return a DataFrame with the explanation of its plan so far.                                                                                                  |
| registry        | Return a `FunctionRegistry` used to plan udf's calls.                                                                                                        |
| schema          | Returns the schema describing the output of this DataFrame in terms of columns returned, where each column has a name, data type, and nullability attribute. |
| to_logical_plan | Return the logical plan represented by this DataFrame.                                                                                                       |

# Expressions

DataFrame methods such as `select` and `filter` accept one or more logical expressions and there are many functions
available for creating logical expressions. These are documented below.

Expressions can be chained together using a fluent-style API:

```rust

> col("a").gt(lit(5)).and(col("b").lt(lit(7)))

Reviewer: ```suggestion
// create the expression `(a > 5) AND (b < 7)`
col("a").gt(lit(5)).and(col("b").lt(lit(7)))
```

```

> ## Identifiers

Reviewer: Since these are not data frame specific, maybe we should put them into a different guide -- and reference that here

Something like `docs/source/user-guide/expressions.md` perhaps

This is just a suggestion, I think it would be fine to do in a follow on PR or never

Author: I will move into a separate page as a follow-on. How would you see users using expressions outside the context of DataFrame usage?

Reviewer: We use them all the time in IOx when we need to make `Exprs` -- for example to create `LogicalPlans` / use the `LogicalPlanBuilder`. I also think they are useful for writing tests when people are writing extensions for DataFusion

| sqrt(x)               | square root                                       |
| tan(x)                | tangent                                           |
| trunc(x)              | truncate toward zero                              |

## Conditional Expressions

| Function | Notes                                                                                                                                                                                                    |
| -------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| coalesce | Returns the first of its arguments that is not null. Null is returned only if all arguments are null. It is often used to substitute a default value for null values when data is retrieved for display. |
| case     | CASE expression. Example: `case(expr).when(expr, expr).when(expr, expr).otherwise(expr).end()`.                                                                                                          |

> | nullif   | Returns a null value if value1 equals value2; otherwise it returns value1. This can be used to perform the inverse operation of the `coalesce` expression.                                               |

Reviewer: ```suggestion
| nullif   | Returns a null value if `value1` equals `value2`; otherwise it returns `value1`. This can be used to perform the inverse operation of the `coalesce` expression.                                               |
```

> ## String Expressions

Reviewer: I think it is ok to leave the `Notes` column blank for now and fill it out going forward

Reviewer: Or maybe we can link to the content in docs/source/user-guide/sql/scalar_functions.md

<!---
  Licensed to the Apache Software Foundation (ASF) under one
  or more contributor license agreements.  See the NOTICE file
  distributed with this work for additional information
  regarding copyright ownership.  The ASF licenses this file
  to you under the Apache License, Version 2.0 (the
  "License"); you may not use this file except in compliance
  with the License.  You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing,
  software distributed under the License is distributed on an
  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
  KIND, either express or implied.  See the License for the
  specific language governing permissions and limitations
  under the License.
-->

# DataFrame API

> DataFrame represents a logical set of rows with the same named columns.
> Similar to a [Pandas DataFrame](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html) or
> [Spark DataFrame](https://spark.apache.org/docs/latest/sql-programming-guide.html)

Reviewer 2: Consider:

```suggestion
DataFrame represents a logical set of rows with the same named columns, similar to a [Pandas DataFrame](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html) or
[Spark DataFrame](https://spark.apache.org/docs/latest/sql-programming-guide.html).
```

> DataFrames are typically created by calling a method on
> `SessionContext`, such as `read_csv` and can then be modified
> by calling the transformation methods, such as `filter`, `select`, `aggregate`, and `limit`
> to build up a query definition.

Reviewer 2: Consider:

```suggestion
DataFrames are typically created by calling a method on
`SessionContext`, such as `read_csv`, and can then be modified
by calling the transformation methods, such as `filter`, `select`, `aggregate`, and `limit`
to build up a query definition.
```

Or perhaps:

```suggestion
DataFrames are typically created by calling a method on
`SessionContext` - such as `read_csv` - and can then be modified
by calling the transformation methods - such as `filter`, `select`, `aggregate`, and `limit` -
to build up a query definition.
```
